### PR TITLE
Hamal 16 snippets

### PR DESCRIPTION
**Pull request by jonsberg**

### CODE REVIEW

File: platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/snippet/SnippetCreateControllerTest.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        val res = createSnippet(
+            ApiCreateSnippetReq(
+                name = SnippetName("TestSnippet"),
+                inputs = SnippetInputs(MapType(mutableMapOf("hamal" to StringType("rocks")))),
+                value = CodeValue("13 + 37")
+            )
+        )
+
+        awaitCompleted(res.reqId)
+        val sn = snippetQueryRepository.get(res.id(::SnippetId))
```
Comment (dominique-dymke) [inline platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/snippet/SnippetCreateControllerTest.kt:28]: Can be inlined - or you dont use with at all

File: platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/snippet/SnippetGetControllerTest.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        with(getSnippetResponse.result(ApiSnippet::class)) {
+            assertThat(id, equalTo(snippetId))
+            assertThat(name, equalTo(SnippetName("TestSnippet")))
+            assertThat(inputs, equalTo(SnippetInputs(MapType(mutableMapOf("hamal" to StringType("rockz"))))))
+            assertThat(value, equalTo(CodeValue("1 + 1")))
+        }
+    }
+
+    @Test
+    fun `get snippet that does not exist`() {
```
Comment (dominique-dymke) [inline platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/snippet/SnippetGetControllerTest.kt:45]:
```suggestion
    fun `Tries to get snippet that does not exist`() {
```

File: platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/snippet/SnippetUpdateControllerTest.kt [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import io.hamal.lib.kua.type.StringType
+import io.hamal.lib.sdk.api.ApiCreateSnippetReq
+import io.hamal.lib.sdk.api.ApiError
+import io.hamal.lib.sdk.api.ApiSubmittedReqWithId
+import org.hamcrest.MatcherAssert.assertThat
+import org.hamcrest.Matchers.equalTo
+
+internal class SnippetUpdateControllerTest : SnippetBaseControllerTest() {
+    @Test
+    fun `Updates snippet that does not exist`() {
```
Comment (dominique-dymke) [inline platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/snippet/SnippetUpdateControllerTest.kt:23]:
```suggestion
    fun `Tries to update snippet that does not exist`() {
```

File: platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/snippet/SnippetUpdateControllerTest.kt [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+        assertThat(getUpdateResponse.statusCode, equalTo(HttpStatusCode.NotFound))
+        require(getUpdateResponse is ErrorHttpResponse) { "request was successful" }
+
+        val error = getUpdateResponse.error(ApiError::class)
+        assertThat(error.message, equalTo("Snippet not found"))
+    }
+
+    @Test
+    fun `Updates snippet`() {
```
Comment (dominique-dymke) [inline platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/snippet/SnippetUpdateControllerTest.kt:42]: Happy case on top and error case afterwards

File: platform/backend/core/src/integrationTest/kotlin/io/hamal/core/req/handler/snippet/CreateSnippetHandlerTest.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import io.hamal.repository.api.submitted_req.SubmittedCreateSnippetReq
+import org.hamcrest.MatcherAssert.assertThat
+import org.hamcrest.Matchers.equalTo
+import org.junit.jupiter.api.Test
+import org.springframework.beans.factory.annotation.Autowired
+
+internal class CreateSnippetHandlerTest : BaseReqHandlerTest() {
+
+    @Autowired
+    private lateinit var testInstance: CreateSnippetHandler
```
Comment (dominique-dymke) [inline platform/backend/core/src/integrationTest/kotlin/io/hamal/core/req/handler/snippet/CreateSnippetHandlerTest.kt:18]: variables at the end of the file

general structure

tests
helper functions 
variables

Reason being, you start reading the file most of the time top to bottom. You usually care about tests and not the variable defiinitions

File: platform/backend/core/src/integrationTest/kotlin/io/hamal/core/req/handler/snippet/UpdateSnippetHandlerTest.kt [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+import io.hamal.repository.api.submitted_req.SubmittedUpdateSnippetReq
+import org.hamcrest.MatcherAssert.assertThat
+import org.hamcrest.Matchers.equalTo
+import org.junit.jupiter.api.Test
+import org.springframework.beans.factory.annotation.Autowired
+
+internal class UpdateSnippetHandlerTest : BaseReqHandlerTest() {
+
+    @Autowired
+    private lateinit var testInstance: UpdateSnippetHandler
```
Comment (dominique-dymke) [inline platform/backend/core/src/integrationTest/kotlin/io/hamal/core/req/handler/snippet/UpdateSnippetHandlerTest.kt:23]: same

File: platform/backend/core/src/integrationTest/kotlin/io/hamal/core/req/handler/snippet/UpdateSnippetHandlerTest.kt [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                    )
+                ),
+                value = CodeValue("1 + 1"),
+                accountId = testAccount.id
+            )
+        )
+
+        testInstance(submittedUpdateSnippetReq)
+
+        val res = snippetQueryRepository.get(SnippetId(123))
```
Comment (dominique-dymke) [inline platform/backend/core/src/integrationTest/kotlin/io/hamal/core/req/handler/snippet/UpdateSnippetHandlerTest.kt:47]: can either be inlined or without with()

File: platform/backend/core/src/main/kotlin/io/hamal/core/adapter/Snippet.kt [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        req: UpdateSnippetReq,
+        responseHandler: (SubmittedReqWithGroupId) -> T
+    ): T {
+        ensureSnippetExists(snippetId)
+        return responseHandler(submitRequest(snippetId, req))
+    }
+
+    private fun ensureSnippetExists(snippetId: SnippetId)  = snippetQueryRepository.get(snippetId)
+
+
```
Comment (dominique-dymke) [inline platform/backend/core/src/main/kotlin/io/hamal/core/adapter/Snippet.kt:65]:
```suggestion
```

File: platform/backend/core/src/main/kotlin/io/hamal/core/req/handler/snippet/CreateSnippetHandler.kt [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+class CreateSnippetHandler(
+    val snippetCmdRepository: SnippetCmdRepository,
+    val eventEmitter: PlatformEventEmitter,
+) : ReqHandler<SubmittedCreateSnippetReq>(SubmittedCreateSnippetReq::class) {
+    override fun invoke(req: SubmittedCreateSnippetReq) {
+        createSnippet(req).also { emitEvent(req.cmdId(), it) }
+    }
+}
+
+    private fun CreateSnippetHandler.createSnippet(req: SubmittedCreateSnippetReq): Snippet {
```
Comment (dominique-dymke) [inline platform/backend/core/src/main/kotlin/io/hamal/core/req/handler/snippet/CreateSnippetHandler.kt:23]: fix indentation

File: platform/backend/repository/api/src/main/kotlin/io/hamal/repository/api/Snippet.kt [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+@Serializable
+data class Snippet(
+    override val id: SnippetId,
+    val groupId: GroupId,
+    val cmdId: CmdId,
+    val name: SnippetName,
+    val inputs: SnippetInputs,
+    val value: CodeValue,
+    val accountId: AccountId
+) : DomainObject<SnippetId>
```
Comment (dominique-dymke) [inline platform/backend/repository/api/src/main/kotlin/io/hamal/repository/api/Snippet.kt:19]:
```suggestion
data class Snippet(
    val cmdId: CmdId,
    override val id: SnippetId,
    val groupId: GroupId,
    val accountId: AccountId,
    val name: SnippetName,
    val inputs: SnippetInputs,
    val value: CodeValue
) : DomainObject<SnippetId>
```

File: platform/backend/repository/api/src/main/kotlin/io/hamal/repository/api/Snippet.kt [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    fun update(snippetId: SnippetId, cmd: UpdateCmd): Snippet
+
+    data class CreateCmd(
+        val id: CmdId,
+        val snippetId: SnippetId,
+        val groupId: GroupId,
+        val inputs: SnippetInputs,
+        val name: SnippetName,
+        val value: CodeValue,
+        val accountId: AccountId
```
Comment (dominique-dymke) [inline platform/backend/repository/api/src/main/kotlin/io/hamal/repository/api/Snippet.kt:34]:
```suggestion
    data class CreateCmd(
        val id: CmdId,
        val snippetId: SnippetId,
        val groupId: GroupId,
         val accountId: AccountId,
        val inputs: SnippetInputs,
        val name: SnippetName,
        val value: CodeValue,
```

File: platform/backend/repository/api/src/main/kotlin/io/hamal/repository/api/submitted_req/Snippet.kt [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+@Serializable
+data class SubmittedCreateSnippetReq(
+    override val reqId: ReqId,
+    override var status: ReqStatus,
+    override val groupId: GroupId,
+    val id: SnippetId,
+    val name: SnippetName,
+    val inputs: SnippetInputs,
+    val value: CodeValue,
+    val accountId: AccountId
```
Comment (dominique-dymke) [inline platform/backend/repository/api/src/main/kotlin/io/hamal/repository/api/submitted_req/Snippet.kt:17]: same

File: platform/backend/repository/record/src/main/kotlin/io/hamal/repository/record/snippet/Entity.kt [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+data class SnippetEntity(
+    override val cmdId: CmdId,
+    override val id: SnippetId,
+    val groupId: GroupId,
+    override val sequence: RecordSequence,
+
+    var name: SnippetName? = null,
+    var inputs: SnippetInputs? = null,
+    var codeValue: CodeValue? = null,
+    var accountId: AccountId? = null
```
Comment (dominique-dymke) [inline platform/backend/repository/record/src/main/kotlin/io/hamal/repository/record/snippet/Entity.kt:19]: move up to groupId  &  account id must not nullable

**Comment by dominique-dymke** [comment]
Please rename accountId of snippet into creatorId - it is still an AccountId class though

File: platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/SnippetTest.kt [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+                    name = SnippetName("TestSnippet"),
+                    inputs = SnippetInputs(
+                        MapType(
+                            mutableMapOf(
+                                "hamal" to StringType("rockz")
+                            )
+                        )
+                    ),
+                    value = CodeValue("1 + 1"),
+                    accountId = AccountId("123")
```
Comment (dominique-dymke) [inline platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/SnippetTest.kt:44]: same same

File: platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/SnippetTest.kt [hunk 15]
```diff
@@ -0,0 +1,349 @@
...
+            }
+            verifyCount(1)
+        }
+
+    }
+
+    @Nested
+    inner class FindTest {
+
+
```
Comment (dominique-dymke) [inline platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/SnippetTest.kt:159]:
```suggestion

```

File: platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/SnippetTest.kt [hunk 16]
```diff
@@ -0,0 +1,349 @@
...
+                groupId = GroupId(1),
+                name = SnippetName("TestSnippet")
+            )
+
+            val exception = assertThrows<NoSuchElementException> {
+                get(SnippetId(111111))
+            }
+            assertThat(exception.message, equalTo("Snippet not found"))
+        }
+    }
```
Comment (dominique-dymke) [inline platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/SnippetTest.kt:192]: This must be part of GetTest - as you are testing the get method not the find method! We need a negative test for find as well in the end we should have 4 tests for get and find

FindTest{
 // we find the snippet
// we dont find the snippet -- null - no exception
}

GetTest{
// we get the snippet
// we are getting an exception
}

File: platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/fixture/MemoryFixture.kt [hunk 17]
```diff
@@ -29,8 +29,9 @@ object MemoryFixture : BaseTestFixture {
         HookRepository::class -> MemoryHookRepository() as REPO
         NamespaceRepository::class -> MemoryNamespaceRepository() as REPO
         ReqRepository::class -> MemoryReqRepository() as REPO
-        StateRepository::class -> MemoryStateRepository() as REPO
```
Comment (dominique-dymke) [inline platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/fixture/MemoryFixture.kt:32]: :+1:

File: platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/fixture/SqliteFixture.kt [hunk 18]
```diff
@@ -76,7 +77,14 @@ object SqliteFixture : BaseTestFixture {
...
 
-        StateRepository::class ->
+        SnippetRepository::class -> SqliteSnippetRepository(
+            SqliteSnippetRepository.Config(createTempDirectory("sqlite_snippet_test"))
+        ) as REPO
+
+
+        StateRepository::class
+
+        ->
```
Comment (dominique-dymke) [inline platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/fixture/SqliteFixture.kt:87]: formatting

File: platform/lib/domain/src/main/kotlin/io/hamal/lib/domain/vo/Snippet.kt [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+
+@Serializable(with = SnippetId.Serializer::class)
+class SnippetId(override val value: SnowflakeId) : DomainId() {
+    constructor(value: Int) : this(SnowflakeId(value.toLong()))
+    constructor(value: String) : this(SnowflakeId(value.toLong(16)))
+
+    internal object Serializer : DomainIdSerializer<SnippetId>(::SnippetId)
+}
+
+@Serializable()
```
Comment (dominique-dymke) [inline platform/lib/domain/src/main/kotlin/io/hamal/lib/domain/vo/Snippet.kt:22]:
```suggestion
@Serializable(with=SnippetName.Serialiser::class)
```

File: platform/runner-extension/std/sys/src/main/resources/sys/extension.lua [hunk 20]
```diff
@@ -105,6 +106,18 @@ function create_extension_factory()
             return internal.req_get(req_id)
         end
 
+        function export.snippet.create(cmd)
+             return internal.snippet_create({
+                            name = cmd.name or nil,
+                            inputs = cmd.inputs or {},
+                            value = cmd.value or ""
+                        })
+        end
```
Comment (dominique-dymke) [inline platform/runner-extension/std/sys/src/main/resources/sys/extension.lua:115]: formatting looks funny

File: platform/testbed/src/integrationTest/resources/admin/snippet/create_get/01.lua [hunk 21]
```diff
@@ -0,0 +1,21 @@
+sys = require('sys')
+
+err, create_snippet_req = sys.snippet.create({
+    name = 'test-snippet',
+    inputs = {},
+    value = [=[40 + 2]=]
```
Comment (dominique-dymke) [inline platform/testbed/src/integrationTest/resources/admin/snippet/create_get/01.lua:6]: LoL what does [=[ do?

File: platform/testbed/src/integrationTest/resources/admin/snippet/create_get/01.lua
Comment (dominique-dymke) [inline platform/testbed/src/integrationTest/resources/admin/snippet/create_get/01.lua:1]: very cool - can you add a snippet update test as well?

File: platform/testbed/src/integrationTest/resources/admin/snippet/create_get/01.lua [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+})
+
+sys.await_completed(create_snippet_req)
+assert(err == nil)
```
Comment (dominique-dymke) [inline platform/testbed/src/integrationTest/resources/admin/snippet/create_get/01.lua:10]:
```suggestion
assert(err == nil)
sys.await_completed(create_snippet_req)
```

I did this wrong in the past myself - we want to make sure that err == ni before we are accessing create_snippet_req. Reason being - create_snippet_req is nil in the event of an error --> null pointer exception

File: platform/testbed/src/integrationTest/resources/admin/snippet/create_get/01.lua (hunk 21 above)
Comment (jonsberg) [reply platform/testbed/src/integrationTest/resources/admin/snippet/create_get/01.lua:6]: [=[  ]=] escapes everything in between, remove it for consistency

File: platform/testbed/src/integrationTest/resources/admin/snippet/update/01.lua [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+
+assert(err == nil)
+sys.await_completed(create_req)
+
+err, update_req = sys.snippet.update(create_req.id, {
+    name = 'update-snippet',
+    inputs = {},
+    value = [[i was updated]]
+})
+
```
Comment (dominique-dymke) [inline platform/testbed/src/integrationTest/resources/admin/snippet/update/01.lua:17]: we have to await update_req complete here as well

**Pull request merged**